### ISSUE TITLE
[Design] Pathways single-controller PD: alternative TransferBackend for #1196

### ISSUE THREAD

**Issue by zorrofox** [sgl-project/sglang-jax#1427]
# [Design] Pathways single-controller PD: alternative transfer backend for #1196

> Sub-issue of #1196 (PD Disaggregation Roadmap). Design doc for review per the contribution guide. PR: `#<PR-NUM>`.

## 1. Motivation

The current `JaxTransferBackend` (#1199) moves KV across pods via `jax.experimental.transfer`, which stages `device → host → NIC → host → device` because TPU lacks device-to-device RDMA across pods. On v6e-1 the staged D2H + p2p pull segments dominate the transfer timeline (~2.5 GB/s/host effective).

**Pathways IFRT proxy already solves the per-host fan-out inside the runtime.** A single Python process sees all devices across N slices; `jax.device_put(array, sharding_on_other_slice)` is lowered to per-host DCN sends without surfacing host buffers to Python. Same hardware (2× v7x-8 slices, DCN not ICI), measured:

| transport | aggregate BW | per-host | payload |
|---|---|---|---|
| `jax.experimental.transfer` (`JaxTransferBackend` primitive) | 9.0 GB/s | 4.5 GB/s | 2048 MiB |
| Pathways IFRT cross-slice `device_put` | **28.9 GB/s** | 14.4 GB/s | 2048 MiB |
| — | — | — | 576 MiB (≈16K-tok MLA KV): **23 ms** |

3.2× is pure software-stack difference; the runtime is the same DCN NIC.

This proposal adds Pathways as an **alternative `TransferBackend`** under the #1199 abstraction (and #1196 roadmap), plus two roadmap items it unblocked along the way: a measured "CO-interference" dimension for PD model selection, and an SWA-aware migrate design (Stage 5).

## 2. Positioning

**Not a parallel architecture.** This slots under the existing PD stack:

- **Reuse** (#1241/#1240/#1368, future Stage 4/6): mini_lb router, bootstrap room, decode admission, PD metrics, P/D ratio policy.
- **Replace** only the #1199 `TransferBackend` layer: instead of `register_pull → pull → release` over `jax.experimental.transfer`, do `gather → stack → device_put → scatter` in one address space.
- **Adds** a single-controller scheduler variant (`_pd_prefill_loop`) that drives N prefill workers + 1 decode worker from one process — #1196 Stage 6 "multi-P single-D" falls out naturally (no inter-process bootstrap).

`--pd-disaggregation=pathways` selects this backend; the existing `--disaggregation-mode={prefill,decode}` multi-process path is untouched.

## 3. Design

```
┌──────────────────────────── Pathways head (CPU node) ───────────────────────────┐
│  scheduler.py                                                                   │
│   ├── tp_workers_p[0..n_p-1]  ModelWorker(mesh=p_mesh_i)   sync prefill         │
│   ├── tp_worker               ModelWorkerClient(mesh=d_mesh) overlap decode     │
│   ├── _pd_prefill_loop[i]  thread: prefill_q → forward(P_i) → gather_to_dmesh   │
│   │                                → ready_q.put((reqs, d_stacked_future))      │
│   └── event_loop_overlap   main:   ready_q → scatter_from_dmesh → running_batch │
│                                                                                 │
│  PathwaysPDKVTransfer (per P slice)                                             │
│    gather_to_dmesh:  P-pool[pages] --gather→ stack[L,npg,...] --device_put→ D   │
│    scatter_from_dmesh: D-pool.kv_buffer = scatter_jit(D-pool, d_pages, stacked) │
└─────────────────────────────────────────────────────────────────────────────────┘
        IFRT proxy ↕ (gRPC)                         IFRT proxy ↕
┌── slice 0..n_p-1 (v7x-N) ──┐            ┌── slice n_p (v7x-N) ──┐
│   prefill forward + KV     │ ──DCN──►   │  decode forward + KV  │
└────────────────────────────┘            └───────────────────────┘
```

**Key invariants**
- `disable_radix_cache=True` (so P-side allocs are page-head aligned → page-level migrate is exact).
- `gather_to_dmesh` runs on the prefill thread and returns a *future* on D-mesh; main-thread `scatter_from_dmesh` consumes it via data-dependency, so cross-slice RecvRefs interleaves with decode forwards instead of stalling (~250 ms otherwise).
- `scatter_jit` donates `kv_buffer`; `_donate_lock` orders the buffer reassignment vs the next decode dispatch.
- `npg` is bucketed to {8,16,32,64} so jit cache size stays O(4).

**Weight load** (single-controller specific): the head process loads once into `_PD_WEIGHT_CACHE` on the first P mesh, then `_pd_remap_sharding` + cross-slice `device_put` populates D in ~24 s instead of re-reading ~700 GB. `SGLANG_MOE_BULK_READ` widens the safetensors bulk-read path (mmap page-fault → 3.05× faster cold load); `SGLANG_MOE_BULK_BLOCK` bounds head-node RSS for >768 GB checkpoints.

**GC**: `gc.freeze()` + raised gen2 threshold at event-loop start. Without it, Pathways IFRT's 100 % `jitted_run_model` C++ cache miss allocates ~60 obj/step → gen2 sweep over ~919 k tracked objects every ~56 s → 2 s GIL hold → Max-ITL spike. With it: Max-ITL 2247 → 269 ms (−88 %).

## 4. Roofline addendum: CO-interference dimension

The N_p/N_d roofline derives the P:D ratio from bandwidth and load shape, and hybrid-attention models look PD-feasible because their KV output is small. **Missing: when the colocated baseline doesn't interfere, PD is net-negative even if transfer is free.**

v7x-32, same hardware (2× 2x2x4 slices vs 1× tp32), same image, IL=4K OL=1K, 2 runs σ<1%:

| model | attn | CO MedITL | CO P99-ITL @rate=0.28 | PD MedITL | PD P99-ITL | PD vs CO MedITL |
|---|---|---|---|---|---|---|
| GLM-5.2 (W8A8) | MLA full | 40.61 | **2162** (blows up) | 35.35 | **82** | **−13 %** |
| GLM-5.2 (FP8) | MLA full | — | — | 36.50 | 73 | — |
| MiMo-V2.5-Pro | 60/70 SWA w=128 | **20.05** | **24** (flat) | 57.82 | 140 | **+188 %** |

GLM's CO P99 explodes because a 4K MLA prefill (~2.2 s TTFT) blocks decode; PD isolates it → P99 −96 %. MiMo's SWA-128 prefill is 469 ms and the CO scheduler hides it → P99 stays 24 ms → halving D's TP for a dedicated P slice just makes decode 2× slower.

**Proposed gate** (cheap, before any PD work on a new model): run CO at 2-3 rates, plot P99-ITL. If it stays flat → PD is the wrong tool for that (model, IL) pair. SWA / linear-attention models at short IL likely fall here; MLA / full-attention models are the PD sweet spot.

## 5. SWA window-only migrate (Stage 5, applies to both backends)

Current `migrate_reqs_p_to_d` moves all `seq_len` tokens for both `full_kv_pool` and `swa_kv_pool`. For SWA the decode side only ever reads the last `window` tokens, so:

- D's effective capacity = `min(full_pool, swa_pool)` → swa pool becomes the bottleneck (observed: D OOM at rate>0.28 on MiMo).
- ~97 % of swa-pool transfer is dead bytes at IL=4K, w=128.

**Design**: split migrate per sub-pool. `full` keeps current behavior. `swa` allocs `ceil(window/page_size)` pages on D and transfers only the last `window` slots from P (page-remapped via `full_to_swa_index_mapping`). The PR ships the dual-sub-pool transfer plumbing (`PathwaysPDKVTransfer.is_swa` + `_swa_pages`); the window-truncated alloc is a follow-up that also benefits the multi-process `KVTransferManager`.

## 6. Shared pain point with `JaxTransferBackend`

`gather/scatter` donate the KV buffer, so they serialize with the next forward via `_donate_lock`. The `JaxTransferBackend` path has the same "gather is on the critical path" issue. Candidate fix (both backends): double-buffer the page indices being scattered so donate happens on a shadow slice while forward reads the live one. Out of scope for this PR; flagging for #1196 Stage 6.

## 7. Supporting fixes bundled in the PR

| file | what | why it can't be split out |
|---|---|---|
| `kernels/_pathways_compat.py` | monkeypatch `_slice_memref` lowering | jax 0.8.1 emits strided-reshape memref that Pathways `parseSourceString` re-verifies and rejects; libtpu in-memory IR skips verify so native is unaffected |
| `fused_moe/v1/kernel.py` | `memory_stats()` fallback 96 GB | IFRT-proxy device raises on `memory_stats()` |
| `weight_utils.py` BULK_BLOCK | per-group `block_until_ready` | async device_put to remote workers lets host shard buffers accumulate to ~963 GB → head-node OOM |
| `jax_utils.py` | dedup `live_arrays()` by buffer ptr | proxy `addressable_shards` creates a new wrapper per access → HBM-used reports 2× → KV pool init computes negative free |
| `tp_worker_overlap_thread.py` | `gc.freeze()` at loop start | §3 GC stall |
| `model_config.py` | deepcopy lru_cached `hf_config` | two ModelConfigs (P+D) mutate the cached object in-place |
| `scheduler_profiler_mixing.py` | `gs://` + `max_num_hosts` profiler args | Pathways worker can only write GCS |

Each is gated by `JAX_PLATFORMS=proxy` or an env var; native multi-process path is byte-identical.

## 8. Known limitations

- No chunked prefill (IL must be ≤ `chunked_prefill_size`). Stage 5 follow-up.
- `dp_size=1` only.
- Pathways runtime image is GCP-only (same closed-surface as `libtpu`, which sglang-jax already depends on). Client side (`pathwaysutils`, CRD, IFRT client) is Apache-2.0; the new code is pure JAX API and unit-tests on CPU mock mesh.
- `pyproject.toml` pins `pathwaysutils<0.1.9` (0.1.9 requires jax≥0.8.3, sglang-jax pins 0.8.1).
- Cold load ~18 min for 78-layer GLM (down from 80 min pre-BULK_READ); D copy ~24 s.

## 9. Benchmarks (v7x-32, same total HW, 2 runs each, σ<1%)

| metric | colocated tp32 | PD p1d1 | Δ |
|---|---|---|---|
| GLM-5.2 W8A8 MedITL @0.19 | 40.61 | 35.35 | −13 % |
| GLM-5.2 W8A8 TTFT @0.19 | 2205 | 1703 | −23 % |
| GLM-5.2 P99-ITL @0.28 | 2162 | 82 | −96 % |
| GLM-5.2 per-chip goodput @P99<200 ms | sat<0.28 | sat>0.28 | ≥1.14× |
| MiMo-V2.5-Pro MedITL @0.19 | 20.05 | 57.82 | +188 % (expected, §4) |

## 10. Test plan

- **CPU unit** (in PR): `XLA_FLAGS=--xla_force_host_platform_device_count=8 JAX_PLATFORMS=cpu pytest test/srt/disaggregation/test_pathways_pd.py` — 8 tests covering mesh split, byte-exact transfer, bucket padding, SWA dual-pool, SWA page mapping.
- **e2e correctness**: GLM-5.1-FP8 78L greedy "What is the capital of France" → `Paris` byte-identical PD vs CO.
- **e2e perf**: reproduce §9 with PathwaysJob CRD `numSlices=2` on v7x 2×(2x2x4) + `--pd-disaggregation=pathways --pd-num-prefill=1`. Deployment yaml in PR description.

## 11. Alternatives considered

- **Single-slice ICI sub-mesh** (ppermute on a `pd` axis): 625 GB/s transfer but hits a libtpu async-collective lowering bug on multi-host device subsets (v6e-specific), and halves TP on both sides so per-chip throughput ≤0.5× CO at decode-heavy ratios. Dropped.
- **Wait for tpu-raiden** (#1196 Stage 5 item): orthogonal — raiden targets the multi-process `TransferBackend`; this backend targets the single-controller niche where one head can see all slices.

**Comment by zorrofox** [comment]
@cjx0709 for your review.

**Comment by cjx0709** [comment]
Already merged
